File: source/RevitDevTool.Test/Extensions/GeometryExtensions.cs
namespace RevitDevTool.Test.Extensions;

[PublicAPI]
public static class GeometryExtensions
{
    public static List<Solid> GetSolids(this Element element)
    {
        var opts = new Options
        {
            ComputeReferences = false,
            IncludeNonVisibleObjects = true
        };

        var solids = new List<Solid>();
        var geometry = element.get_Geometry(opts);
        if (geometry == null) return solids;

        foreach (var geoObj in geometry)
        {
            switch (geoObj)
            {
                case Solid solid when solid.Volume != 0:
                    solids.Add(solid);
                    break;
                case GeometryInstance geoInstance:
                    foreach (var instGeoObj in geoInstance.GetInstanceGeometry())
                    {
                        if (instGeoObj is Solid instSolid && instSolid.Volume != 0)
                        {
                            solids.Add(instSolid);
                        }
                    }
                    break;
            }
        }

        return solids;
    }

    public static List<Mesh> GetMeshes(this Element element)
    {
        var opts = new Options
        {
            ComputeReferences = false,
            IncludeNonVisibleObjects = true
        };
        var meshes = new List<Mesh>();
        var geometry = element.get_Geometry(opts);
        if (geometry == null) return meshes;
        foreach (var geoObj in geometry)
        {
            switch (geoObj)
            {
                case Mesh mesh:
                    meshes.Add(mesh);
                    break;
                case GeometryInstance geoInstance:
                    foreach (var instGeoObj in geoInstance.GetInstanceGeometry())
                    {
                        if (instGeoObj is Mesh instMesh)
                        {
                            meshes.Add(instMesh);
                        }
                    }
                    break;
            }
        }
        return meshes;
    }

    public static List<Face> GetFaces(this Solid solid)
    {
        return solid.Faces.Cast<Face>().ToList();
    }

    public static List<XYZ> GetVertices(this Mesh mesh)
    {
        return mesh.Vertices.ToList();
    }

    public static List<XYZ> GetVertices(this Solid solid)
    {
        return solid.Edges.Cast<Edge>().SelectMany(edge => edge.Tessellate()).ToList();
    }

    public static List<XYZ> GetVertices(this GeometryObject geometryObject)
    {
        return geometryObject switch
        {
            Solid solid => solid.GetVertices(),
            Mesh mesh => mesh.GetVertices(),
            _ => []
        };
    }

    public static List<Edge> GetEdges(this Solid solid)
    {
        return solid.Edges.Cast<Edge>().ToList();
    }

    public static List<Curve> GetCurves(this Solid solid)
    {
        return solid.Edges.Cast<Edge>().Select(edge => edge.AsCurve()).ToList();
    }

    public static List<Edge> GetEdges(this Face face)
    {
        var edgeArrays = face.EdgeLoops.Cast<EdgeArray>();
        return edgeArrays.SelectMany(edgeLoop => edgeLoop.Cast<Edge>()).ToList();
    }

    public static List<Curve> GetCurves(this Face face)
    {
        var curveLoops = face.GetEdgesAsCurveLoops();
        return curveLoops?.SelectMany(loop => loop).ToList() ?? [];
    }
}

File: samples/FSharpDemo/Extensions/GeometryExtensions.fs
module RevitDevTool.FSharpDemo.Extensions.GeometryExtensions

open System.Linq
open Autodesk.Revit.DB

[<AutoOpen>]
module GeometryExtensions =

    let private geometryOptions () =
        new Options(ComputeReferences = false, IncludeNonVisibleObjects = true)

    let getSolids (element: Element) =
        let opts = geometryOptions ()
        let geometry = element.get_Geometry(opts)
        if geometry = null then
            []
        else
            [ for geoObj in geometry do
                match geoObj with
                | :? Solid as solid when solid.Volume <> 0.0 -> yield solid
                | :? GeometryInstance as geoInstance ->
                    for instGeoObj in geoInstance.GetInstanceGeometry() do
                        match instGeoObj with
                        | :? Solid as instSolid when instSolid.Volume <> 0.0 -> yield instSolid
                        | _ -> ()
                | _ -> () ]

    let getMeshes (element: Element) =
        let opts = geometryOptions ()
        let geometry = element.get_Geometry(opts)
        if geometry = null then
            []
        else
            [ for geoObj in geometry do
                match geoObj with
                | :? Mesh as mesh -> yield mesh
                | :? GeometryInstance as geoInstance ->
                    for instGeoObj in geoInstance.GetInstanceGeometry() do
                        match instGeoObj with
                        | :? Mesh as instMesh -> yield instMesh
                        | _ -> ()
                | _ -> () ]

    let getFacesOfSolid (solid: Solid) =
        solid.Faces.Cast<Face>() |> Seq.toList

    let getVerticesOfMesh (mesh: Mesh) =
        mesh.Vertices |> Seq.toList

    let getVerticesOfSolid (solid: Solid) =
        solid.Edges.Cast<Edge>()
        |> Seq.collect (fun edge -> edge.Tessellate())
        |> Seq.toList

    let getVerticesOfGeometryObject (geoObj: GeometryObject) =
        match geoObj with
        | :? Solid as solid -> getVerticesOfSolid solid
        | :? Mesh as mesh -> getVerticesOfMesh mesh
        | _ -> []

    let getEdgesOfSolid (solid: Solid) =
        solid.Edges.Cast<Edge>() |> Seq.toList

    let getCurvesOfSolid (solid: Solid) =
        solid.Edges.Cast<Edge>()
        |> Seq.map (fun edge -> edge.AsCurve())
        |> Seq.toList

    let getEdgesOfFace (face: Face) =
        face.EdgeLoops.Cast<EdgeArray>()
        |> Seq.collect (fun edgeLoop -> edgeLoop.Cast<Edge>())
        |> Seq.toList

    let getCurvesOfFace (face: Face) =
        let curveLoops = face.GetEdgesAsCurveLoops()
        if curveLoops = null then
            []
        else
            curveLoops |> Seq.collect id |> Seq.toList

